pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,262 @@
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	sharing "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	"github.com/liqotech/liqo/pkg/discovery"
+	fcutils "github.com/liqotech/liqo/pkg/utils/foreignCluster"
+)
+
+type peeringCache struct {
+	peeringInfo sync.Map
+	ready       bool
+}
+
+/**
+ * PeeringCache methods
+ */
+
+// CacheRefresher is a wrapper funcion that receives a ShadowPodValidator and starts an PollImmediateInfinite timer to periodically refresh the cache.
+func CacheRefresher(spv *Validator, interval time.Duration) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		return wait.PollImmediateInfiniteWithContext(ctx, interval, spv.refreshCache)
+	}
+}
+
+func (spv *Validator) initializeCache(ctx context.Context) (err error) {
+	resourceOfferList := sharing.ResourceOfferList{}
+	if err := spv.client.List(ctx, &resourceOfferList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	cacherefreshlog.Info("Cache initialization started")
+	for i := range resourceOfferList.Items {
+		ro := &resourceOfferList.Items[i]
+		clusterID := ro.Labels[discovery.ClusterIDLabel]
+		clusterName := retrieveClusterName(ctx, spv.client, clusterID)
+		if clusterID == "" {
+			return fmt.Errorf("ResourceOffer %s has no cluster id", clusterName)
+		}
+		cacherefreshlog.Info(fmt.Sprintf("Generating PeeringInfo in cache for corresponding ResourceOffer %s", clusterID))
+		pi := createPeeringInfo(clusterID, clusterName, ro.Spec.ResourceQuota.Hard)
+
+		// Get the List of shadow pods running on the cluster with a given clusterID
+		shadowPodList, err := spv.getShadowPodListByClusterID(ctx, clusterID)
+		if err != nil {
+			return err
+		}
+
+		alignExistingShadowPods(shadowPodList, pi)
+
+		spv.peeringCache.peeringInfo.Store(clusterID, pi)
+	}
+	cacherefreshlog.Info(fmt.Sprintf("Cache initialization completed. Found %d peered clusters", len(resourceOfferList.Items)))
+	spv.peeringCache.ready = true
+	return nil
+}
+
+func (spv *Validator) refreshCache(ctx context.Context) (done bool, err error) {
+	if !spv.peeringCache.ready {
+		err = spv.initializeCache(ctx)
+		if err != nil {
+			cacherefreshlog.Info(err.Error())
+		}
+		return false, nil
+	}
+	resourceOfferList := sharing.ResourceOfferList{}
+	if err := spv.client.List(ctx, &resourceOfferList, &client.ListOptions{}); err != nil {
+		cacherefreshlog.Info(err.Error())
+		return false, nil
+	}
+	cacherefreshlog.Info("Cache refresh started")
+	// Cycling on all recorded peering in Cache
+	spv.peeringCache.peeringInfo.Range(
+		func(key, value interface{}) bool {
+			pi := value.(*peeringInfo)
+			clusterID := key.(string)
+			// Get the List of shadow pods running on the cluster
+			shadowPodList, err := spv.getShadowPodListByClusterID(ctx, clusterID)
+			if err != nil {
+				cacherefreshlog.Info(err.Error())
+				return false
+			}
```

Comment:
Also in this case I would continue with the next element, instead of blocking completely.